rust/otap-dataflow/crates/otap/src/grpc.rs

```diff
@@ -0,0 +1,213 @@
+        request: Request<tonic::Streaming<BatchArrowRecords>>,
+    ) -> Result<Response<Self::ArrowMetricsStream>, Status> {
+        let mut input_stream = request.into_inner();
+        let (tx, rx) = tokio::sync::mpsc::channel(self.channel_size);
+        let effect_handler_clone = self.effect_handler.clone();
+
+        // Provide client a stream to listen to
+        let output = ReceiverStream::new(rx);
+
+        // write to the channel
+        _ = tokio::spawn(async move {
+            // Process messages until stream ends or error occurs
+            while let Ok(Some(batch)) = input_stream.message().await {
+                // accept the batch data and handle output response
+                if accept_data(OTAPData::ArrowMetrics, batch, &effect_handler_clone, &tx)
+                    .await
+                    .is_err()
+                {
+                    // end loop if error occurs
+                    break;
+                }
+            }
+        });
+
+        Ok(Response::new(Box::pin(output) as Self::ArrowMetricsStream))
+    }
+}
+
+#[tonic::async_trait]
+impl ArrowTracesService for ArrowTracesServiceImpl {
+    type ArrowTracesStream =
+        Pin<Box<dyn Stream<Item = Result<BatchStatus, Status>> + Send + 'static>>;
+    async fn arrow_traces(
+        &self,
+        request: Request<tonic::Streaming<BatchArrowRecords>>,
+    ) -> Result<Response<Self::ArrowTracesStream>, Status> {
+        let mut input_stream = request.into_inner();
+        let (tx, rx) = tokio::sync::mpsc::channel(self.channel_size);
+        let effect_handler_clone = self.effect_handler.clone();
+
+        // create a stream to output result to
+        let output = ReceiverStream::new(rx);
+
+        // write to the channel
+        _ = tokio::spawn(async move {
+            // Process messages until stream ends or error occurs
+            while let Ok(Some(batch)) = input_stream.message().await {
+                // accept the batch data and handle output response
+                if accept_data(OTAPData::ArrowTraces, batch, &effect_handler_clone, &tx)
+                    .await
+                    .is_err()
+                {
+                    // end loop if error occurs
+                    break;
+                }
+            }
+        });
+
+        Ok(Response::new(Box::pin(output) as Self::ArrowTracesStream))
+    }
+}
+
+/// handles sending the data down the pipeline via effect_handler and generating the approriate response
+async fn accept_data<OTAPDataType>(
+    otap_data: OTAPDataType,
+    batch: BatchArrowRecords,
+    effect_handler: &shared::EffectHandler<OTAPData>,
+    tx: &tokio::sync::mpsc::Sender<Result<BatchStatus, Status>>,
+) -> Result<(), ()>
+where
+    OTAPDataType: Fn(BatchArrowRecords) -> OTAPData,
+{
+    let batch_id = batch.batch_id;
+    let status_result = match effect_handler.send_message(otap_data(batch)).await {
+        Ok(_) => (StatusCode::Ok, "Successfully received".to_string()),
```

Comment:
Should the message here be "Successfully sent" considering this is a result of `send_message` operation?
```suggestion
        Ok(_) => (StatusCode::Ok, "Successfully sent".to_string()),
```

Reply:
Hmmm, this remark makes me think we might want to evaluate terminology used for sending/receiving, for receiving/exporting, consuming/producing, etc.